libgadget/slotsmanager.cu

```diff
@@ -0,0 +1,751 @@
+ *              If you are using this function on the output of slots_split_particle, it should be false.
+ */
+int
+slots_convert(int parent, int ptype, int placement, struct part_manager_type * pman, struct slots_manager_type * sman)
+{
+    /* Explicitly mark old slot as garbage */
+    int oldtype = pman->Base[parent].Type;
+    int oldPI = pman->Base[parent].PI;
+    if(oldPI >= 0 && SLOTS_ENABLED(oldtype, sman))
+        BASESLOT_PI(oldPI, oldtype, sman)->ReverseLink = pman->MaxPart + 100;
+
+    /* Make a new slot */
+    if(SLOTS_ENABLED(ptype, sman)) {
+        int newPI = placement;
+        /* If enabled, allocate a new slot for secondary data */
+        if(placement < 0)
+            newPI = atomic_fetch_and_add_64(&sman->info[ptype].size, 1);
+
+        /* There is no clear way to safely grow the slots during this, because the memory may be deep in the heap.*/
+        if(newPI >= sman->info[ptype].maxsize) {
+            endrun(1, "Tried to use non-allocated slot %d (> %ld)\n", newPI, sman->info[ptype].maxsize);
+        }
+        slots_connect_new_slot(parent, newPI, ptype, pman, sman);
+    }
+    /* Type changed after slot updated */
+    pman->Base[parent].Type = ptype;
+    return parent;
+}
+
+/* This will split a new particle out from an existing one, conserving mass.
+ * The type is the same, and the slot PI on the new particle is set to -1.
+ * You should call slots_convert on the child afterwards to create a new slot.
+ *
+ * Assumes the particle is protected by locks in threaded environments.
+ *
+ * The Generation of parent is incremented.
+ * The child carries the incremented generation number.
+ * The ID of the child is modified, with the new generation number set
+ * at the highest 8 bits.
+ *
+ * The new particle's index is returned.
+ *
+ * Its mass and ptype can be then adjusted using slots_convert.
+ * The 'new particle' event is emitted.
+ */
+int
+slots_split_particle(int parent, double childmass, struct part_manager_type * pman)
+{
+    int64_t child = atomic_fetch_and_add_64(&pman->NumPart, 1);
+
+    if(child >= pman->MaxPart)
+        endrun(8888, "Tried to spawn: NumPart=%ld MaxPart=%ld. Sorry, no space left.\n", child, pman->MaxPart);
+
+    pman->Base[parent].Generation++;
+    uint64_t g = pman->Base[parent].Generation;
+    pman->Base[child] = pman->Base[parent];
+
+    /* Change the child ID according to the generation. */
+    pman->Base[child].ID = (pman->Base[parent].ID & 0x00ffffffffffffffL) + (g << 56L);
+    if(g >= (1 << (64-56L)))
+        endrun(1, "Particle %d (ID: %ld) generated too many particles: generation %ld wrapped.\n", parent, pman->Base[parent].ID, g);
+
+    pman->Base[child].Mass = childmass;
+    pman->Base[parent].Mass -= childmass;
+
+    /* Invalidate the slot of the child. Call slots_convert soon afterwards! */
+    pman->Base[child].PI = -1;
+
+    return child;
+}
+
+/* Remove garbage particles, holes in sph chunk, and holes in bh buffer.
```

Comment:
You can remove this function and the functions it calls from this file: no need to port it to gpu.